udev/99-stratisd.rules

```diff
@@ -0,0 +1 @@
+ENV{DM_NAME}=="stratis-1-[0-9a-f]*-thin-fs-[0-9a-f]*", PROGRAM="/usr/bin/stratis_uuids_to_names.py $env{DM_NAME}", SYMLINK+="stratis/%c{1}/%c{2}"
```

Comment:
How does using the ```+=``` operator result in the old symlink, on a rename of, say, a filesystem, being removed? It sure looks like it works, so what is the implicit rule that causes this to happen?

Reply:
My understanding here is that it is because of the type of synthetic event that stratisd is generating. I have not tested what would happen if we generated an `add` event, but the `change` event certainly seems to work. This made sense to me as the best option to try when I wrote it because a `change` event is responsible for updating an existing entry. As such, if udev has any awareness of which symlinks are associated with a rule (and it appears from the behavior that it may), `+=` would mean that we don't clobber existing symlinks generated from other rules (like `/dev/mapper` symlinks for example) and udev would be able to clean up the old symlink associated with this rule if it differs from the new one and then add the new one that we've specified. I'm not sure if this is exactly how it works, but it doesn't seem like an unreasonable guess based on the behavior and the limited digging I've done into udev rules.

Reply:
> My understanding here is that it is because of the type of synthetic event that stratisd is generating. I have not tested what would happen if we generated an `add` event, but the `change` event certainly seems to work. This made sense to me as the best option to try when I wrote it because a `change` event is responsible for updating an existing entry. As such, if udev has any awareness of which symlinks are associated with a rule (and it appears from the behavior that it may), `+=` would mean that we don't clobber existing symlinks generated from other rules (like `/dev/mapper` symlinks for example) and udev would be able to clean up the old symlink associated with this rule if it differs from the new one and then add the new one that we've specified. I'm not sure if this is exactly how it works, but it doesn't seem like an unreasonable guess based on the behavior and the limited digging I've done into udev rules.

Exactly...it seems like we ought to use ```+=``` in order not to clobber the symlinks established by other rules. Note that there is another assignment operator, which implies final, that is, no other rule can assign a new value to the symlink. So that all makes sense.

But, is there some implicit udev semantics that says a single rule can only ever make one symlink? Apparently not, as there is a syntax for specifying multiple symlinks at one time. Maybe the semantics for a rule is that every time the rule gets run, all the effects of previously running the rule get erased?

My current hypothesis is that the udev database keeps track of what rule caused what thing to be set, and when the rule is triggered it undoes the previous effects of the rule. I have no evidence for this, beyond observation.

Reply:
I'll leave this as unresolved until we have a better understanding of this.